In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as s
import seaborn as sns
from IPython.display import Markdown, display
import math

In [8]:
def markdown(input):
    display(Markdown(input))

In [9]:
data_path = "../data/anggur.csv"
df = pd.read_csv(data_path)
df.columns = ["fixed_acidity", "volatile_acidity", "citric_acid", "residual_sugar", "chlorides", "free_sulfur_dioxide", 
              "total_sulfur dioxide", "density", "pH", "sulphates", "alcohol", "quality"]
df

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0             5.90            0.4451       0.1813        2.049401   0.070574   
1             8.40            0.5768       0.2099        3.109590   0.101681   
2             7.54            0.5918       0.3248        3.673744   0.072416   
3             5.39            0.4201       0.3131        3.371815   0.072755   
4             6.51            0.5675       0.1940        4.404723   0.066379   
..             ...               ...          ...             ...        ...   
995           7.96            0.6046       0.2662        1.592048   0.057555   
996           8.48            0.4080       0.2227        0.681955   0.051627   
997           6.11            0.4841       0.3720        2.377267   0.042806   
998           7.76            0.3590       0.3208        4.294486   0.098276   
999           5.87            0.5214       0.1883        2.179490   0.052923   

     free_sulfur_dioxide  total_sulfur dioxide  density    pH  sulphates  \
0              16.593818                 42.27   0.9982  3.27       0.71   
1              22.555519                 16.01   0.9960  3.35       0.57   
2               9.316866                 35.52   0.9990  3.31       0.64   
3              18.212300                 41.97   0.9945  3.34       0.55   
4               9.360591                 46.27   0.9925  3.27       0.45   
..                   ...                   ...      ...   ...        ...   
995            14.892445                 44.61   0.9975  3.35       0.54   
996            23.548965                 25.83   0.9972  3.41       0.46   
997            21.624585                 48.75   0.9928  3.23       0.55   
998            12.746186                 44.53   0.9952  3.30       0.66   
999            16.203864                 24.37   0.9983  3.29       0.70   

     alcohol  quality  
0       8.64        7  
1      10.03        8  
2       9.23        8  
3      14.07        9  
4      11.49        8  
..       ...      ...  
995    10.41        8  
996     9.91        8  
997     9.94        7  
998     9.76        8  
999    10.17        7  

[1000 rows x 12 columns]

##### Nomor 1
Menulis deskripsi statistika (Descriptive Statistics) dari semua kolom pada data yang bersifat
numerik, terdiri dari mean, median, modus, standar deviasi, variansi, range, nilai minimum,
maksimum, kuartil, IQR, skewness dan kurtosis. Boleh juga ditambahkan deskripsi lain.

In [10]:
from tabulate import tabulate
def describe(column_name): 
    table = {}
    temp = {}
    df1 = df[column_name]
    temp["Mean"] = df1.mean()
    temp["Median"] = df1.median()
    temp["Mode"] = df1.mode()[0]
    temp["Deviation"] = df1.std()
    temp["Variation"] = df1.var()
    temp["Range"] = df1.max() - df1.min()
    temp["Min"] = df1.min()
    temp["Max"] = df1.max()
    temp["1st Quartile"] = df1.quantile(0.25)
    temp["2nd Quartile"] = df1.quantile(0.5)
    temp["3rd Quartile"] = df1.quantile(0.75)
    temp["IQR"] = df1.quantile(0.75) - df1.quantile(0.25)  
    temp["Skew"] = df1.skew()
    temp["Kurtosis"] = df1.kurtosis()
    table[column_name] = temp
    print(tabulate([k, v] for k, v in temp.items()))

    

In [14]:
for column in list(df) :
    markdown(f"#### {column}")
    describe(column) 

#### fixed_acidity

------------  ----------
Mean           7.15253
Median         7.15
Mode           6.54
Deviation      1.2016
Variation      1.44384
Range          8.17
Min            3.32
Max           11.49
1st Quartile   6.3775
2nd Quartile   7.15
3rd Quartile   8
IQR            1.6225
Skew          -0.0288786
Kurtosis      -0.0192921
------------  ----------


#### volatile_acidity

------------  -----------
Mean           0.520839
Median         0.52485
Mode           0.5546
Deviation      0.0958483
Variation      0.00918689
Range          0.6652
Min            0.1399
Max            0.8051
1st Quartile   0.4561
2nd Quartile   0.52485
3rd Quartile   0.585375
IQR            0.129275
Skew          -0.197699
Kurtosis       0.161853
------------  -----------


#### citric_acid

------------  -----------
Mean           0.270517
Median         0.2722
Mode           0.3019
Deviation      0.0490984
Variation      0.00241065
Range          0.2929
Min            0.1167
Max            0.4096
1st Quartile   0.2378
2nd Quartile   0.2722
3rd Quartile   0.302325
IQR            0.064525
Skew          -0.0455761
Kurtosis      -0.104679
------------  -----------


#### residual_sugar

------------  ----------
Mean           2.5671
Median         2.51943
Mode           0.0325545
Deviation      0.987915
Variation      0.975977
Range          5.5182
Min            0.0325545
Max            5.55075
1st Quartile   1.89633
2nd Quartile   2.51943
3rd Quartile   3.22087
IQR            1.32454
Skew           0.132638
Kurtosis      -0.04298
------------  ----------


#### chlorides

------------  ------------
Mean           0.0811952
Median         0.0821669
Mode           0.0151224
Deviation      0.0201106
Variation      0.000404438
Range          0.125635
Min            0.0151224
Max            0.140758
1st Quartile   0.0665736
2nd Quartile   0.0821669
3rd Quartile   0.0953115
IQR            0.0287379
Skew          -0.0513193
Kurtosis      -0.246508
------------  ------------


#### free_sulfur_dioxide

------------  -----------
Mean          14.9077
Median        14.8603
Mode           0.194679
Deviation      4.8881
Variation     23.8935
Range         27.2678
Min            0.194679
Max           27.4625
1st Quartile  11.4267
2nd Quartile  14.8603
3rd Quartile  18.3131
IQR            6.88638
Skew           0.00713042
Kurtosis      -0.364964
------------  -----------


#### total_sulfur dioxide

------------  ----------
Mean          40.2902
Median        40.19
Mode          35.2
Deviation      9.96577
Variation     99.3165
Range         66.81
Min            3.15
Max           69.96
1st Quartile  33.785
2nd Quartile  40.19
3rd Quartile  47.0225
IQR           13.2375
Skew          -0.02406
Kurtosis       0.0639498
------------  ----------


#### density

------------  ------------
Mean           0.995925
Median         0.996
Mode           0.9959
Deviation      0.00202018
Variation      4.08113e-06
Range          0.0138
Min            0.9888
Max            1.0026
1st Quartile   0.9946
2nd Quartile   0.996
3rd Quartile   0.9972
IQR            0.0026
Skew          -0.0768828
Kurtosis       0.0163656
------------  ------------


#### pH

------------  ---------
Mean          3.30361
Median        3.3
Mode          3.34
Deviation     0.104875
Variation     0.0109989
Range         0.74
Min           2.97
Max           3.71
1st Quartile  3.23
2nd Quartile  3.3
3rd Quartile  3.37
IQR           0.14
Skew          0.147673
Kurtosis      0.0809096
------------  ---------


#### sulphates

------------  ---------
Mean          0.59839
Median        0.595
Mode          0.59
Deviation     0.100819
Variation     0.0101645
Range         0.67
Min           0.29
Max           0.96
1st Quartile  0.53
2nd Quartile  0.595
3rd Quartile  0.67
IQR           0.14
Skew          0.149199
Kurtosis      0.0648193
------------  ---------


#### alcohol

------------  ----------
Mean          10.5923
Median        10.61
Mode           9.86
Deviation      1.51071
Variation      2.28223
Range          8.99
Min            6.03
Max           15.02
1st Quartile   9.56
2nd Quartile  10.61
3rd Quartile  11.6225
IQR            2.0625
Skew          -0.0189914
Kurtosis      -0.131732
------------  ----------


#### quality

------------  ----------
Mean           7.958
Median         8
Mode           8
Deviation      0.902802
Variation      0.815051
Range          5
Min            5
Max           10
1st Quartile   7
2nd Quartile   8
3rd Quartile   9
IQR            2
Skew          -0.0890541
Kurtosis       0.108291
------------  ----------


##### Nomor 2
Membuat Visualisasi plot distribusi, dalam bentuk histogram dan boxplot untuk setiap kolom numerik. Berikan uraian penjelasan kondisi setiap kolom berdasarkan kedua plot
tersebut.

#### Nomor 3 
Menentukan setiap kolom numerik berdistribusi normal atau tidak. Gunakan normality test
yang dikaitkan dengan histogram plot.

##### Nomor 4
Melakukan test hipotesis 1 sampel,
+ Nilai rata-rata pH di atas 3.29?
+ Nilai rata-rata Residual Sugar tidak sama dengan 2.50?
+ Nilai rata-rata 150 baris pertama kolom sulphates bukan 0.65?
+ Nilai rata-rata total sulfur dioxide di bawah 35?
+ Proporsi nilai total sulfur dioxide yang lebih dari 40, adalah tidak sama dengan 50% ?

##### Nomor 5
Melakukan test hipotesis 2 sampel,
+ Data kolom fixed acidity dibagi 2 sama rata: bagian awal dan bagian akhir kolom.
Benarkah rata-rata kedua bagian tersebut sama?
+ Data kolom chlorides dibagi 2 sama rata: bagian awal dan bagian akhir kolom. Benarkah
rata-rata bagian awal lebih besar daripada bagian akhir sebesar 0.001?
+ Benarkah rata-rata sampel 25 baris pertama kolom Volatile Acidity sama dengan
rata-rata 25 baris pertama kolom Sulphates ?
+ Bagian awal kolom residual sugar memiliki variansi yang sama dengan bagian akhirnya?
+ Proporsi nilai setengah bagian awal alcohol yang lebih dari 7, adalah lebih besar daripada, proporsi nilai yang sama di setengah bagian akhir alcohol?